### PR TITLE
Add gradient for clip

### PR DESCRIPTION
**Pull request by caisq**
This change is [Reviewable](https://reviewable.io/reviews/pair-code/deeplearnjs/644)

### CODE REVIEW

File: src/math/unary_ops.ts [hunk 1]
```diff
@@ -121,8 +122,14 @@ export class Ops {
         (min <= max),
         `Error in clip: min (${min}) must be` +
             `less than or equal to max (${max}).`);
-    return ENV.engine.executeKernel('Clip', {inputs: {x}, args: {min, max}}) as
-        T;
+    return ENV.engine.executeKernel(
+        'Clip', {inputs: {x}, args: {min, max}}, (dy: T, y: T) => {
+      return {
+          x: () => dy.where(
```
Comment (ManrajGrover) [inline src/math/unary_ops.ts]: Gradient for `clip` is not defined for x = min and x = max. We should add a TODO here to fix this once it is defined. Reference: #606

File: src/math/unary_ops.ts [hunk 2]
```diff
@@ -121,8 +122,14 @@ export class Ops {
         (min <= max),
         `Error in clip: min (${min}) must be` +
             `less than or equal to max (${max}).`);
-    return ENV.engine.executeKernel('Clip', {inputs: {x}, args: {min, max}}) as
-        T;
+    return ENV.engine.executeKernel(
```
Comment (ManrajGrover) [inline src/math/unary_ops.ts]: `dl.where` currently does not support `Scalar` values. We should throw an error for this.

File: src/math/unary_ops_test.ts [hunk 3]
```diff
@@ -1309,6 +1309,18 @@ import {Array1D, Array2D, Scalar} from './ndarray';
       };
       expect(f).toThrowError();
     });
+
```
Comment (ManrajGrover) [inline src/math/unary_ops_test.ts]: Let's add tests for `Scalars` to check considering it is not supported yet.

**Comment by caisq** [comment]
oops, I need to pull and rebase.

**Comment by nsthorat** [comment]
Review status: 0 of 2 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[src/math/unary_ops.ts, line 125 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/644#-L4asXvnfc1509SSs16T-r1-125:-L4foGJpzeu09uWqRFB0:bxfz60) ([raw file](https://github.com/pair-code/deeplearnjs/blob/ede5139e436885fcb61a77c0af2aeb1edc71acf2/src/math/unary_ops.ts#L125)):*
<details><summary><i>Previously, ManrajGrover (Manraj Singh) wrote…</i></summary><blockquote>

`dl.where` currently does not support `Scalar` values. We should throw an error for this.
</blockquote></details>

Any idea why this is? It should be easy to fix right?

https://github.com/PAIR-code/deeplearnjs/blob/master/src/math/logical_ops.ts#L100

---

*[src/math/unary_ops.ts, line 129 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/644#-L4fnvQruF8tL5IyefIO:-L4fnvQruF8tL5IyefIP:b-xcz9c8) ([raw file](https://github.com/pair-code/deeplearnjs/blob/ede5139e436885fcb61a77c0af2aeb1edc71acf2/src/math/unary_ops.ts#L129)):*
> ```XML
>       return {
>           x: () => dy.where(
>               x.greater(Scalar.new(min)).logicalAnd(x.less(Scalar.new(max))),
> ```

import "scalar" from "ops" and call that function here instead of "Scalar.new", "Scalar.new" is deprecated.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pair-code/deeplearnjs/644)*

**Comment by ManrajGrover** [comment]
@nsthorat 

> Any idea why this is? It should be easy to fix right?
> 
> https://github.com/PAIR-code/deeplearnjs/blob/master/src/math/logical_ops.ts#L100

This happens because `dl.where` internally calls `getCoordsDataType()` which does not support rank 0. We should make it use `int` for both rank 0 and rank 1.

**Comment by ManrajGrover** [comment]
Have opened an issue for this #665 and fix is in #666

File: src/math/unary_ops.ts (hunk 2 above)
Comment (caisq) [reply src/math/unary_ops.ts]: Done.

File: src/math/unary_ops.ts (hunk 1 above)
Comment (caisq) [reply src/math/unary_ops.ts]: Done.

File: src/math/unary_ops_test.ts (hunk 3 above)
Comment (caisq) [reply src/math/unary_ops_test.ts]: Since you are implementing it soon, I decided to not add an explicit unit test for that for now.

**Comment by nsthorat** [comment]
@caisq can you merge with master? @ManrajGrover fixed the scalar issue.

**Comment by caisq** [comment]
@nsthorat Done.

**Comment by caisq** [comment]
PTAL @nsthorat @ManrajGrover ?

**Comment by dsmilkov** [comment]
:lgtm_strong:

---

Review status: 0 of 2 files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/pair-code/deeplearnjs/644#-:-L4krtFcoXSRBtyLL0h0:b-jbsf8z)*

**Comment by dsmilkov** [comment]
Thanks @caisq . All comments were addressed, so I merged!

**Pull request merged**